### DISCUSSION TITLE
How to reduce serverless function size?

### DISCUSSION THREAD

**Discussion by ramiro-sena** [vercel/community#699] [Help] [answered]
Hi,
I'm currently trying to deploy an very basic image processing API as python runtime **Serverless Function** into Vercel. But I get the following error:
> Error: The Serverless Function "server/api.py" is 90.64mb which exceeds the maximum size limit of 50mb. Learn More: https://vercel.link/serverless-function-size

The source Git Repo for this project can be accessed by this url: https://github.com/ramiro-sena/angle_finder/tree/935c4b65e62e824971fe9995ec34062ef9ec7797

I thought the error could have something to do with the imports, so I ran this python file to check the size of all imports I'm using on this project `module_size.py`
```
import os, sys
dir_size = lambda dirpath:sum([(sum([os.path.getsize(os.path.join(root, f)) for f in files]) + sum([dir_size(os.path.join(root, d)) for d in dirs])) for root, files, dirs in os.walk(dirpath)])
def module_size(module_name):
    for p in sys.path:
        if os.path.exists(os.path.join(p, module_name)+'.py'):
            return f"{os.path.getsize(os.path.join(p, module_name)+'.py')/1024.0} KB"
        if os.path.exists(os.path.join(p, module_name)):            
            return f'{dir_size(os.path.join(p, module_name))/1024.0} KB'
    return 'Module not installed.'

modules = ['numpy', 'io', 'starlette', 'uvicorn', 'fastapi', 'cv2']

for m in modules:
  print(f"{m} ===> {module_size(m)}")

```

### OUTPUT: 
> PS C:\Users\ramir\angle-finder-test> python module_size.py 
numpy ===> 528.0 KB
aiofiles ===> 12.0 KB
starlette ===> 16.0 KB
uvicorn ===> 48.0 KB
fastapi ===> 36.0 KB
cv2 ===> 12.0 KB

Looking at [documentation](https://vercel.com/docs/concepts/limits/overview#serverless-function-size) I found that I could use  `includeFiles` and `excludeFiles`. But I have no Idea on how to use.

I would be grateful if someone could give me some instruction or sugestion on how to reduce the Serverles Function size.

**Comment by amyegan** [comment, marked as answer]
Hi @ramiro-sena. Here's the [config options for functions](https://vercel.com/docs/project-configuration#project-configuration/functions). Selecting which files to include or exclude will depend on your project and which packages you're including. I believe `numpy` is the what's causing the problem for you. We have another suggestion from [an old  discussion thread](https://github.com/vercel/vercel/discussions/4709#discussioncomment-30402) suggesting the `public` directory should be excluded like
```
{
  "functions": {
    "api/**/*.py": {
      "excludeFiles": "public/**"
    }
  }
}
```
That could be a good place to start. But if that isn't working for you, this Stack Overflow thread on [reducing Python serverless function size](https://stackoverflow.com/questions/46887794/reduce-size-of-serverless-deploy-package) might help you decide which files to exclude.
